### PR TITLE
simplifier pass: apply subroutine from function & fix return_type of findloc

### PR DESCRIPTION
**Pull request by gxyd**

### CODE REVIEW

**Comment by gxyd** [comment]
I'm moving it to draft, as I first want to see if the CI passes or not. I'll want to think more closely about the changes.

File: src/libasr/pass/intrinsic_array_function_registry.h [hunk 1]
```diff
@@ -3374,7 +3374,7 @@ namespace FindLoc {
     ASR::expr_t *mask = args[3];
     ASR::expr_t *back = args[5];
     if (overload_id == 1) {
-        body.push_back(al, b.Assignment(result, b.i_t(0, return_type)));
+        body.push_back(al, b.Assignment(result, b.i_t(0, ASRUtils::type_get_past_array(return_type))));
```
Comment (gxyd) [inline src/libasr/pass/intrinsic_array_function_registry.h:3377]: This change is definitely a correct one.

File: src/libasr/pass/intrinsic_function.cpp [hunk 2]
```diff
@@ -164,7 +164,8 @@ class ReplaceFunctionCallReturningArray: public ASR::BaseExprReplacer<ReplaceFun
 
     void replace_FunctionCall(ASR::FunctionCall_t* x) {
         ASR::symbol_t* x_m_name = ASRUtils::symbol_get_past_external(x->m_name);
-        if( func2intrinsicid.find(x_m_name) == func2intrinsicid.end() ) {
+        if( func2intrinsicid.find(x_m_name) == func2intrinsicid.end() ||
+            !ASRUtils::IntrinsicArrayFunctionRegistry::handle_dim(func2intrinsicid[x_m_name])) {
```
Comment (gxyd) [inline src/libasr/pass/intrinsic_function.cpp:168]: we should be able to get rid of this change and the `subroutine_from_function` pass applied again. I just have to figure out the bug present in `FindLoc` which isn't allowing that.

**Comment by gxyd** [comment]
I'm more happy with this change in simplifier_pass than main.

**Comment by gxyd** [comment]
I noticed that `matrix_01_transpose` failed sometimes in another PR as well, and when I pushed another commit it randomly succeeded.

For me locally on macOS, I've never seen it fail though.

**Comment by gxyd** [comment]
Pasting the partial CI log says, in case we try to re-run the failing CI job :
```
llvm_wasm_emcc    =   6.75 sec*proc (1060 tests)
mlir              =   0.01 sec*proc (1 test)
Errors while running CTest
wasm              =   1.28 sec*proc (220 tests)
x86               =   0.10 sec*proc (19 tests)

Total Test time (real) =   1.98 sec

The following tests FAILED:
	133 - matrix_01_transpose (SEGFAULT)
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-llvm
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-llvm2
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-llvm_rtlib
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-c
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-cpp
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-x86
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-wasm
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-gfortran
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-llvmImplicit
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-llvmStackArray
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-fortran
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-c_nopragma
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-llvm_nopragma
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-llvm_wasm
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-llvm_wasm_emcc
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-llvm_omp
+ rm -rf /home/runner/work/lfortran/lfortran/integration_tests/test-mlir
+ mkdir /home/runner/work/lfortran/lfortran/integration_tests/test-llvm
+ FC=lfortran cmake -DLFORTRAN_BACKEND=llvm -DFAST=no -DCURRENT_BINARY_DIR=/home/runner/work/lfortran/lfortran/integration_tests/test-llvm -S /home/runner/work/lfortran/lfortran/integration_tests -B /home/runner/work/lfortran/lfortran/integration_tests/test-llvm
```

**Comment by gxyd** [comment]
Ok, I re-ran it the second time, and the job "Check debug build" job succeeded now.

The failing test (in the initial run) was independent of this change. I couldn't the find CI job where `matrix_01_transpose` earlier randomly failed in a PR though, I'll try to find one tomorrow maybe.

**Comment by kmr-srbh** [comment]
The test `matrix_01_transpose` had encountered a segfault in my last PR also. I think it's strange that it showing up often, though everything seems to be fine. I will try to take a look at the test and the internal implementations to see if there might be something flaky.

File: src/libasr/pass/intrinsic_array_function_registry.h [hunk 3]
```diff
@@ -5503,18 +5505,6 @@ namespace IntrinsicArrayFunctionRegistry {
...
-        // Dim argument is already handled for the following
-        if( id == IntrinsicArrayFunctions::Shape  ||
-            id == IntrinsicArrayFunctions::MaxLoc ||
-            id == IntrinsicArrayFunctions::MinLoc ||
-            id == IntrinsicArrayFunctions::FindLoc ) {
-            return false;
-        } else {
-            return true;
-        }
-    }
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_array_function_registry.h:5516]: we need this, rest all is good.

**Review by Pranavchiku — APPROVED** [review]
The PR is good, we just need to revert removal of `handle_dim`.

**Comment by gxyd** [comment]
> The PR is good, we just need to revert removal of `handle_dim`.

Alright, I've reverted the removal of `handle_dim`.

**Review by Pranavchiku — APPROVED** [review]

**Comment by gxyd** [comment]
Thanks for the review @Pranavchiku . I've enabled "squash & merge" on this, as it made more sense for this PR.

**Review by HarshitaKalani — COMMENTED** [review]
Thanks @gxyd for the changes. They look good.

**Pull request merged**